### PR TITLE
feature/add-ssl certs-verify-resource

### PR DESCRIPTION
**Pull request by jschoewe**
Added resource for verifying, revoking, and renewing vault ssl certificates

### CODE REVIEW

File: lib/puppet/type/vault_cert.rb [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    desc 'IP Subject Alternative Names'
+  end
+
+  newparam(:vault_server) do
+    desc 'Hostname of the Vault server'
+  end
+
+  newparam(:vault_scheme) do
+    desc 'Hostname of the Vault server'
+    defaultto('http')
```
Comment (nmaludy) [inline lib/puppet/type/vault_cert.rb]: default to https (secure by default)
```suggestion
    defaultto('https')
```

File: lib/puppet/type/vault_cert.rb [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    validate do |value|
+      path = Pathname.new(value)
+      unless path.absolute?
+        raise ArgumentError, "Path must be absolute: #{path}"
+      end
+    end
+  end
+
+  newparam(:priv_key_path) do
+    desc 'The path to the private key'
```
Comment (nmaludy) [inline lib/puppet/type/vault_cert.rb]: default to the same as the cert_path ?

File: lib/puppet/type/vault_cert.rb [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    validate do |value|
+      path = Pathname.new(value)
+      unless path.absolute?
+        raise ArgumentError, "Path must be absolute: #{path}"
+      end
+    end
+  end
```
Comment (nmaludy) [inline lib/puppet/type/vault_cert.rb:84]: do we need a path to write the CA file as well?

File: lib/puppet/type/vault_cert.rb [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+  newparam(:password) do
+    desc 'The optional password for the private key'
+  end
+
+  newparam(:auth_type) do
+    desc 'authentication type of the private key'
```
Comment (nmaludy) [inline lib/puppet/type/vault_cert.rb:110]: what is a good default here?

File: lib/puppet/type/vault_cert.rb [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    validate do |value|
+      acceptable_values = ['dsa', 'rsa', 'ec']
+      unless acceptable_values.include? value.downcase
+        raise ArgumentError, "auth_type must be one of: #{acceptable_values.join(', ')}"
+      end
+    end
+  end
+  
+  newparam(:ttl_hours_remaining) do
+    desc 'Number of hours remaining before the cert needs to be renewed'
```
Comment (nmaludy) [inline lib/puppet/type/vault_cert.rb]: may an improvement of 
```suggestion
    desc 'Re-generate and replace the certificate this many hours before it expires.'
```

File: lib/puppet/type/vault_cert.rb [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    desc 'authentication type of the private key'
+    validate do |value|
+      acceptable_values = ['dsa', 'rsa', 'ec']
+      unless acceptable_values.include? value.downcase
+        raise ArgumentError, "auth_type must be one of: #{acceptable_values.join(', ')}"
+      end
+    end
+  end
+  
+  newparam(:ttl_hours_remaining) do
```
Comment (nmaludy) [inline lib/puppet/type/vault_cert.rb]: maybe rename to like `regenerate_ttl` or `regen_ttl` to take the same naming standard as `cert_ttl` below?

File: lib/puppet/type/vault_cert.rb [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+  newparam(:cert_ttl) do
+    desc 'TTL to give the new cert'
+    defaultto('720h')
+  end
+
+  newparam(:sans) do
+    desc 'IP Subject Alternative Names'
+  end
+
+  newparam(:vault_server) do
```
Comment (nmaludy) [inline lib/puppet/type/vault_cert.rb]:
```suggestion
  newparam(:api_server) do
```

File: lib/puppet/type/vault_cert.rb [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    desc 'Hostname of the Vault server'
```
Comment (nmaludy) [inline lib/puppet/type/vault_cert.rb]:
```suggestion
    desc 'Hostname/IP Address of the Vault server'
```

File: lib/puppet/type/vault_cert.rb [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+  end
+
+  newparam(:vault_scheme) do
+    desc 'Hostname of the Vault server'
```
Comment (nmaludy) [inline lib/puppet/type/vault_cert.rb]:
```suggestion
    desc 'Communication scheme/transport for the API ('http', 'https')'
```

File: lib/puppet/type/vault_cert.rb [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+  newparam(:sans) do
+    desc 'IP Subject Alternative Names'
+  end
+
+  newparam(:vault_server) do
+    desc 'Hostname of the Vault server'
+  end
+
+  newparam(:vault_scheme) do
```
Comment (nmaludy) [inline lib/puppet/type/vault_cert.rb]:
```suggestion
  newparam(:api_scheme) do
```

File: lib/puppet/type/vault_cert.rb [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+    desc 'Hostname of the Vault server'
+    defaultto('http')
+  end
+
+  newparam(:vault_port) do
```
Comment (nmaludy) [inline lib/puppet/type/vault_cert.rb]:
```suggestion
  newparam(:api_port) do
```

File: lib/puppet/type/vault_cert.rb [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    desc 'Hostname of the Vault server'
```
Comment (nmaludy) [inline lib/puppet/type/vault_cert.rb]:
```suggestion
    desc 'Port for communicating with the Vault server'
```

File: lib/puppet/provider/vault_cert/openssl.rb [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+    # Get the directory from the given key path
+    key_dir = File.dirname(resource[:priv_key_path])
+    # Save the new private key in the tls directory on the client
+    key_path = File.join(key_dir, key_name)
+    File.open(key_path, 'w') do |f|
+      f.write(cert['data']['private_key'])
+    end
+  end
+
+  # TODO: Save a copy of the newly generated cert on the vault server
```
Comment (nmaludy) [inline lib/puppet/provider/vault_cert/openssl.rb]: is this TODO still valid?

**Review by nmaludy — CHANGES_REQUESTED** [review]
General comment: I need to setup travis builds so we can make sure this thing is correct from a linting perspective.

File: lib/puppet/provider/vault_cert/openssl.rb (hunk 13 above)
Comment (jschoewe) [reply lib/puppet/provider/vault_cert/openssl.rb]: No , I'll remove it since it isn't necessary

File: lib/puppet/type/vault_cert.rb [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+  end
+
+  newparam(:owner) do
+    desc 'Owner to assign for the new cert and key'
+    defaultto('root')
+  end
+
+  newparam(:group) do
+    desc 'Group to assign for the new cert and key'
+    defaultto('root')
```
Comment (nmaludy) [inline lib/puppet/type/vault_cert.rb:137]: @jschoewe need to add an `alt_names` parameter here too for named subject-alternative-names that are not IP addresses, example (myservice.domain.tld,myservice)

File: lib/puppet/provider/vault_cert/openssl.rb [hunk 15]
```diff
@@ -0,0 +1,283 @@
...
+    headers = {
+      'X-Vault-Token' => resource[:api_token],
+    }
+
+    payload = {
+      serial_number: serial_number,
+    }
+
+    response = send_request('POST', api_path, payload, headers)
+    return JSON.parse(response.body) if response.body
```
Comment (nmaludy) [inline lib/puppet/provider/vault_cert/openssl.rb]: looks like this is copy/pasted a bunch

any reason it's not just put in the `send_request` function?

File: lib/puppet/provider/vault_cert/openssl.rb [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+  end
+
+  # Save an openssl PKey object into the global priv_key var
+  def private_key_get
+    return @priv_key unless @priv_key.nil?
+    priv_key_path = File.join(resource[:priv_key_dir], resource[:priv_key_name])
+    @priv_key = if Pathname.new(priv_key_path).exist?
+                  file = File.read(priv_key_path)
+                  auth_type = resource[:auth_type].downcase
+                  if auth_type == 'dsa'
```
Comment (nmaludy) [inline lib/puppet/provider/vault_cert/openssl.rb]: could replace if/else with `case/when` statement

File: lib/puppet/provider/vault_cert/openssl.rb [hunk 17]
```diff
@@ -0,0 +1,283 @@
...
+                else
+                  false
+                end
+  end
+
+  # Read the serial number from the certificate, convert it to base 16, and add colons
+  def cert_serial_get
+    cert = certificate_get
+    # Get the serial number from the cert object
+    serial_number = cert.serial.to_s(16)
```
Comment (nmaludy) [inline lib/puppet/provider/vault_cert/openssl.rb:201]: what is the 16 here?

**Review by nmaludy — CHANGES_REQUESTED** [review]
Couple really small things

**Review by nmaludy — APPROVED** [review]

**Pull request merged**